Speed up depth average postprocessor

gassmoeller opened the pull request:
This began as a small optimization project and ended up as nearly complete rewrite of the lateral averaging functionality. All external interfaces are kept, and all functionality should work as before, but by reducing duplication the depth average postprocessor is now nearly ten times faster than before. Instead of n loops over all cells for n properties, lateral averaging now performs 1 loop for n properties (by providing n functors that are all executed for each cell). The `compute_average` function no longer takes a `FUNCTOR` template argument, but instead each functor is derived from a common base class, which makes creating a vector of functors much simpler. I am quite happy with the current state so I would like to ask for a review. In particular I would like to ask for comments about the position of the base class `FunctorBase`. I would like to hide it from the header file, but I could not forward declare its type.

I have another improvement lined up, but that one changes the functionality by reducing the number of quadrature points if possible, and I will open that as a separate PR.

bangerth commented on source/simulator/lateral_averaging.cc:
```diff
@@ -448,17 +250,279 @@ namespace aspect
+            ExcMessage ("To call this function, you need to request a positive "
+                        "number of depth slices."));
+    Assert (functors.size() == values.size(),
+            ExcMessage ("To call this function, you need supply as many data "
```
-> need TO  supply

bangerth commented on source/simulator/lateral_averaging.cc:
```diff
@@ -448,17 +250,279 @@ namespace aspect
+    // this yields 10^dim quadrature points evenly distributed in the interior of the cell.
+    // We avoid points on the faces, as they would be counted more than once.
+    const QIterated<dim> quadrature_formula (QMidpoint<1>(),
+                                             10);
```
Is this what we had before? Ten seems kind of a lot. If it's still slow, there would be an easy way to reduce the computational complexity.

gassmoeller replied:
Yes, this is what I meant in my initial comment. I thought about adaptively computing the number of quadrature points based on the largest cell diameter and the size of the depth slices to make sure we use the lowest possible resolution while still having at least one quadrature point per depth slice. Unfortunately this decreases the accuracy of the postprocessor (because the volume fraction that we add to the wrong depth slice increases). @jdannberg preference was to have a slower but more accurate postprocessor. I am not sure about this, we could also make that an input argument.

bangerth replied:
Yes, that's reasonable. After your 10x speedup, does the depth averaging still show up in your profiles?

gassmoeller replied:
It is still 3 times as expensive as visualization, in my example model that is 11% of total runtime, if you run it every timestep. However, I suspect this is not longer crucial as most people will run it at most every 20 timesteps or so. In other words not terribly concerning right now. I thought about better ways to do this, but could not come up with one. In theory a `QAnisotropic` would save a lot of time, since we only need a high resolution in depth direction, but I think we can not rely on the orientation of the cells. Anyway, maybe I just stop here and say it is good enough.

bangerth replied:
Yes, good enough. If someone complains, then we'll drop the number of integration points to 8 in each direction and gain a factor of 2 :-)

bangerth commented on source/simulator/lateral_averaging.cc:
```diff
@@ -448,17 +250,279 @@ namespace aspect
+      {
+        functors[i]->setup(quadrature_formula.size());
+        if (functors[i]->need_material_properties())
+          functors_need_material_output = true;
```
I look forward to the day where we can use `std::reduce` or `std::transform_reduce` for this kind of thing -- but that's only in C++17 :-(
http://en.cppreference.com/w/cpp/algorithm/reduce
http://en.cppreference.com/w/cpp/algorithm/transform_reduce

bangerth commented on source/simulator/lateral_averaging.cc:
```diff
@@ -448,17 +250,279 @@ namespace aspect
+      }
+  }
+
+  template <int dim>
```
here and below: 3 empty lines between functions

bangerth commented on source/simulator/lateral_averaging.cc:
```diff
@@ -448,17 +250,279 @@ namespace aspect
+  }
+
+  template <int dim>
+  void LateralAveraging<dim>::get_composition_averages( const unsigned int c, std::vector<double> &values) const
```
break arguments onto separate lines

bangerth commented on source/simulator/lateral_averaging.cc:
```diff
@@ -448,17 +250,279 @@ namespace aspect
+    std::vector<std::vector<double> > vector_values(1,values);
+    std::vector<std::string> property_name(1,"temperature");
+    get_averages(property_name,vector_values);
+    values.swap(vector_values[0]);
```
if you return the data from `get_averages` as mentioned earlier, you can just do
```
  values = get_averages (property_name);
```

bangerth commented on source/simulator/lateral_averaging.cc:
```diff
@@ -448,17 +250,279 @@ namespace aspect
+  }
+
+  template <int dim>
+  void LateralAveraging<dim>::get_averages(std::vector<std::string> property_names,
```
the first argument should be a `const` reference

gassmoeller commented:
I addressed all comments. I kept the Functor operator as a reference argument (this is for internal use, and in one of the inner loops over all cells), but changed the other functions to return by value, which made the code much nicer.

gassmoeller commented:
Thanks!
